File: android/src/main/kotlin/com/wifi/improv/RpcResultParser.kt
/*
 * Added by MrLiuYunPing on 2026-03-18 for improv_wifi_flutter.
 * This parser accumulates fragmented RPC result notifications and decodes
 * complete Improv BLE result packets, logging checksum mismatches without
 * dropping otherwise valid string payloads.
 */

package com.wifi.improv

internal class RpcResultParser {
    private val buffer = mutableListOf<Byte>()

    fun reset() {
        buffer.clear()
    }

    fun append(chunk: ByteArray): List<List<String>> {
        if (chunk.isEmpty()) {
            return emptyList()
        }

        buffer.addAll(chunk.toList())
        val packets = mutableListOf<List<String>>()

        while (buffer.size >= 3) {
            val payloadLength = buffer[1].toUByte().toInt()
            val packetLength = payloadLength + 3
            if (buffer.size < packetLength) {
                break
            }

            val packet = ByteArray(packetLength) { index -> buffer[index] }
            repeat(packetLength) {
                buffer.removeAt(0)
            }

            decodePacket(packet)?.let(packets::add)
        }

        return packets
    }

    private fun decodePacket(packet: ByteArray): List<String>? {
        if (packet.size < 3) {
            return null
        }

        val payloadLength = packet[1].toUByte().toInt()
        if (packet.size != payloadLength + 3) {
            return null
        }

        val expectedChecksum = calculateChecksum(packet.copyOf(packet.size - 1))
        val actualChecksum = packet.last().toUByte()
        if (expectedChecksum != actualChecksum) {
            // Some devices return a valid URL payload with a checksum that
            // doesn't match the published algorithm. Prefer surfacing the
            // decoded RPC strings instead of dropping the whole result.
        }

        val strings = mutableListOf<String>()
        var currentIndex = 2
        val payloadEnd = packet.size - 1

        while (currentIndex < payloadEnd) {
            val stringLength = packet[currentIndex].toUByte().toInt()
            currentIndex++
            if (currentIndex + stringLength > payloadEnd) {
                return null
            }

            val string = try {
                packet.decodeToString(
                    startIndex = currentIndex,
                    endIndex = currentIndex + stringLength,
                    throwOnInvalidSequence = true
                )
            } catch (_: Exception) {
                return null
            }

            currentIndex += stringLength
            strings += string
        }

        return strings
    }

    private fun calculateChecksum(data: ByteArray): UByte {
        var checksum = 0.toUByte()
        for (byte in data) {
            checksum = (checksum + byte.toUByte()).toUByte()
        }
        return checksum
    }
}

File: ios/Classes/ImprovSDK/RpcResultParser.swift
import Foundation

final class RpcResultParser {
    private var buffer: [UInt8] = []

    func reset() {
        buffer.removeAll()
    }

    func append(_ chunk: Data) -> [[String]] {
        guard !chunk.isEmpty else { return [] }

        buffer.append(contentsOf: chunk)
        var packets: [[String]] = []

        while buffer.count >= 3 {
            let payloadLength = Int(buffer[1])
            let packetLength = payloadLength + 3
            guard buffer.count >= packetLength else {
                break
            }

            let packet = Array(buffer.prefix(packetLength))
            buffer.removeFirst(packetLength)

            if let decoded = decodePacket(packet) {
                packets.append(decoded)
            }
        }

        return packets
    }

    private func decodePacket(_ packet: [UInt8]) -> [String]? {
        guard packet.count >= 3 else { return nil }

        let payloadLength = Int(packet[1])
        guard packet.count == payloadLength + 3 else { return nil }

        let body = Array(packet.dropLast())
        let expectedChecksum = calculateChecksum(body)
        let actualChecksum = packet[packet.count - 1]
        guard expectedChecksum == actualChecksum else { return nil }

        var strings: [String] = []
        var currentIndex = 2
        let payloadEnd = packet.count - 1

        while currentIndex < payloadEnd {
            let stringLength = Int(packet[currentIndex])
            currentIndex += 1

            guard currentIndex + stringLength <= payloadEnd else { return nil }
            let data = Data(packet[currentIndex..<(currentIndex + stringLength)])
            currentIndex += stringLength

            guard let string = String(data: data, encoding: .utf8) else { return nil }
            strings.append(string)
        }

        return strings
    }

    private func calculateChecksum(_ data: [UInt8]) -> UInt8 {
        var checksum: UInt8 = 0
        for byte in data {
            checksum = (checksum &+ byte) & 255
        }
        return checksum
    }
}
